dali/operators/imgcodec/decoder.h

```diff
@@ -0,0 +1,214 @@
+#include "dali/operators/imgcodec/operator_utils.h"
+#include "dali/operators/image/crop/crop_attr.h"
+#include "dali/operators/generic/slice/slice_attr.h"
+#include "dali/operators/image/crop/random_crop_attr.h"
+#include "dali/imgcodec/image_decoder_interfaces.h"
+
+#ifndef DALI_OPERATORS_IMGCODEC_DECODER_H_
+#define DALI_OPERATORS_IMGCODEC_DECODER_H_
+
+namespace dali {
+namespace imgcodec {
+
+template <typename Backend>
+class DecoderBase : public Operator<Backend> {
+ public:
+  ~DecoderBase() override = default;
+
+ protected:
+  explicit DecoderBase(const OpSpec &spec) : Operator<Backend>(spec) {
+    opts_.format = spec.GetArgument<DALIImageType>("output_type");
+    opts_.dtype = spec.GetArgument<DALIDataType>("dtype");
+    GetDecoderSpecificArguments(spec);
+  }
+
+  virtual void SetupRoiGenerator(const OpSpec &spec, const workspace_t<Backend> &ws) {}
+
+  virtual ROI GetRoi(const OpSpec &spec, const workspace_t<Backend> &ws, std::size_t data_idx,
+                     TensorShape<> shape) {
+    return {};
+  }
+
+  virtual int GetDeviceId() = 0;
+
+  template<typename T>
+  void GetDecoderSpecificArgument(const OpSpec &spec, const std::string &name,
+                                  std::function<bool(const T&)> validator = [](T x){return true;}) {
+    if (spec.HasArgument(name)) {
+      auto value = spec.GetArgument<T>(name);
+      if (!validator(value)) {
+        DALI_FAIL(make_string("Invalid value for decoder-specific parameter ", name));
+      }
+      decoder_params_[name] = value;
+    }
+  }
+
+  void GetDecoderSpecificArguments(const OpSpec &spec) {
+    GetDecoderSpecificArgument<uint64_t>(spec, "hybrid_huffman_threshold");
+    GetDecoderSpecificArgument<size_t>(spec, "device_memory_padding");
+    GetDecoderSpecificArgument<size_t>(spec, "host_memory_padding");
+    GetDecoderSpecificArgument<size_t>(spec, "device_memory_padding_jpeg2k");
+    GetDecoderSpecificArgument<size_t>(spec, "host_memory_padding_jpeg2k");
+    GetDecoderSpecificArgument<float>(spec, "hw_decoder_load");
+    GetDecoderSpecificArgument<size_t>(spec, "preallocate_width_hint");
+    GetDecoderSpecificArgument<size_t>(spec, "preallocate_height_hint");
+    GetDecoderSpecificArgument<bool>(spec, "memory_stats");
+    GetDecoderSpecificArgument<bool>(spec, "use_fast_idct");
+  }
+
+  bool CanInferOutputs() const override { return true; }
+
+  void SetupShapes(const OpSpec &spec, const workspace_t<Backend> &ws,
+                    std::vector<OutputDesc> &output_descs, ThreadPool& tp) {
+    output_descs.resize(1);
+    auto &input = ws.template Input<CPUBackend>(0);
+    int nsamples = input.num_samples();
+    rois_.resize(nsamples);
+    SetupRoiGenerator(spec, ws);
+    TensorListShape<> shapes;
+    BatchVector<ImageSource> srcs;
+    BatchVector<ImageSource *> src_ptrs;
+    shapes.resize(nsamples);
+    srcs.resize(nsamples);
+    src_ptrs.resize(nsamples);
+
+    for (int i = 0; i < shapes.size(); i++) {
```

Comment:
In this loop, you're duplicating what CalculateOutputShapes does - you could just populate the source pointers and that's it.